---
title: "AT4"
author: "Tasfia Tasnim, Tanjila Tabassum, Razieh Fathinezhad)"
date: "2024-10-12"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:


## This chunk contains the codes for:
1. Downloading gene_expression.tsv
2. Reading the tsv file
3. Showing the table of the first six genes

```{r}
URL="https://raw.githubusercontent.com/ghazkha/Assessment4/refs/heads/main/gene_expression.tsv"
download.file(URL,destfile="gene_expression.tsv")
expression <- read.table(URL,stringsAsFactors = FALSE, header=TRUE)
expression
firstsixgenes <- expression[1:6,]
firstsixgenes

```

## This chunk contains the codes for:
1. Creating a column which is the mean of other columns
```{r}
expression$GTEX.1117F.0226.SM.5GZZ7 <- as.numeric(expression$GTEX.1117F.0226.SM.5GZZ7)
expression$GTEX.1117F.0426.SM.5EGHI <- as.numeric(expression$GTEX.1117F.0426.SM.5EGHI)
expression$GTEX.1117F.0526.SM.5EGHJ <- as.numeric(expression$GTEX.1117F.0526.SM.5EGHJ)

expression$columnmeans <- rowMeans(expression[,2:4])
head(expression)

```

## This chunk contains the codes for:
1. Determining the 10 genes with highest expressed mean.
```{r}
expression_sorted <- expression[order(-expression$columnmeans), ]
top_10_genes <- head(expression_sorted, 10)
top_10_genes
```

## This chunk contains the code for:
1. Determining the Number of genes mean <10
```{r}
lessexpressed <- subset(expression, columnmeans < 10)
lessexpressed

```

## This chunk contains the code for:
1. Creating histogram of mean values
```{r}
hist(expression$columnmeans,xlab="gene expression",main="mean expression")
```

## This chunk contains the code for:
1. Downloading growth_data.csv file
2. Reading the csv file
3. Showing the name of columns
```{r}
URL="https://raw.githubusercontent.com/ghazkha/Assessment4/refs/heads/main/growth_data.csv"
download.file(URL,destfile="growth_data.csv")
growth <- read.table(URL,sep=",",header=TRUE)
growth
```

## This chunk contains code for:
1. Calculating the mean and standard deviation of start and end study.
```{r}
growth_numeric <- growth[-1,]
growth_numeric$Circumf_2005_cm <- as.numeric(growth_numeric$Circumf_2005_cm)
growth_numeric$Circumf_2010_cm <- as.numeric(growth_numeric$Circumf_2010_cm)
growth_numeric$Circumf_2015_cm <- as.numeric(growth_numeric$Circumf_2015_cm)
growth_numeric$Circumf_2020_cm <- as.numeric(growth_numeric$Circumf_2020_cm)
colMeans(growth_numeric[,3, drop=FALSE])
colMeans(growth_numeric[,6, drop=FALSE])
sd(growth_numeric[,3])
sd(growth_numeric[,6])
```
## This chunk contains code for:
1. Boxplot of start and end of both sites
```{r}
TC2005 <- growth_numeric[,grep("Circumf_2005_cm",colnames(growth_numeric))]
TC2020 <- growth_numeric[,grep("Circumf_2020_cm",colnames(growth_numeric))]
boxplot(TC2005,TC2020,
        ylab="circumference",
        main="growth",
        names = c("2005","2020"))
```

## This chunk contains code for:
1. Calculating the mean growth
```{r}
northeast <- subset(growth_numeric,Site == "northeast")
southwest <- subset(growth_numeric,Site == "southwest")

mean(northeast$Difference <- northeast$Circumf_2020_cm - northeast$Circumf_2010_cm)
mean(southwest$Difference <- southwest$Circumf_2020_cm - southwest$Circumf_2010_cm)
```

## This chunk contains code for:
1. Estimating the p value by executing t.test
```{r}
growth_numeric$growth_10_year <- growth_numeric$Circumf_2020_cm - growth_numeric$Circumf_2010_cm
growth_northeast <- subset(growth_numeric, Site == "northeast")
growth_southwest <- subset(growth_numeric, Site == "southwest")
t_test_result <- t.test(growth_northeast$growth_10_year, growth_southwest$growth_10_year)
t_test_result

```



## This chunk contains code for:
1. installing packages
```{r}
library(R.utils)
library("seqinr")
```

## This chunk contains codes for:
1. Downloading the E.coli and Mycobacterium genomes
2. Reading the files
```{r}
URL="https://ftp.ensemblgenomes.ebi.ac.uk/pub/bacteria/release-59/fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655_gca_000005845/cds/Escherichia_coli_str_k_12_substr_mg1655_gca_000005845.ASM584v2.cds.all.fa.gz"
download.file(URL,destfile="ecoli_cds.fa.gz")
gunzip("ecoli_cds.fa.gz")
cds <- seqinr::read.fasta("ecoli_cds.fa")
URL = "https://ftp.ensemblgenomes.ebi.ac.uk/pub/bacteria/release-59/fasta/bacteria_4_collection/mycobacterium_tuberculosis_gca_001318445/cds/Mycobacterium_tuberculosis_gca_001318445.6505_5_10.cds.all.fa.gz"
download.file(URL,destfile="mycobacterium_cds.fa.gz")
gunzip("mycobacterium_cds.fa.gz")
cds_mycobacterium <- seqinr::read.fasta("mycobacterium_cds.fa")

```
## This chunk contains codes for:
1. Counting the lengths
2. Creating a table of the length
```{r}
length(cds)
length(cds_mycobacterium)
table <- data.frame(
  E.coli = 4239,
  Mycobacterium = 4741
)
print(table)


```
## This chunk contains code for:
1. Calculating total Coding DNA
2. Creating a table of Total Coding DNA
```{r}
head(summary(cds))
head(summary(cds_mycobacterium))

len <- as.numeric(summary(cds)[,1])
lenm <- as.numeric(summary(cds_mycobacterium)[,1])
sum(len)
sum(lenm)
table <- data.frame(
  E.coli = 3978528,
  Mycobacterium = 3076728
)
print(table)
```

## This chunk contains codes for:
1. Creating a boxplot
2. Determining the mean and media 
```{r}
par(mfrow=c(1,2)) 
boxplot(len,ylab="sequence length (bp)",main="E.coli")
boxplot(lenm,ylab="sequence length (bp)",main="Meycobacterium")
mean(len)
mean(lenm)
median(len)
median(lenm)

```


This chunk contains code for:
1. Creating barplots for nucleotide and amino acid
2. Calculating the frequency of DNA bases and total protein
```{r}
E.coliDNA <- unlist(cds)
dna_composition_e.coli <- count(E.coliDNA,1)
MycobacteriumDNA <- unlist(cds_mycobacterium)
dna_composition_mycobacterium <- count(MycobacteriumDNA,1)
par(mfrow=c(1,2))
barplot(dna_composition_e.coli,main="E.coli")
barplot(dna_composition_mycobacterium,main="Mycobacterium")

prot <- lapply(cds, translate)
aa <- unique(prot[[2]])
aa <- aa[aa != "*"]
length(aa)
protm <- lapply(cds_mycobacterium, translate)
am <- unique(protm[[2]])
am <- am[am != "*"]
length(am)
E.coliProt <- unlist (prot)
prot_composition_e.coli <- count(E.coliProt,wordsize=1,alphabet=aa)
MycobacteriumProt <- unlist (protm)
prot_composition_mycobacterium <- count(MycobacteriumProt,wordsize=1,alphabet=am)
par(mfrow=c(1,2))
barplot(prot_composition_e.coli,main="E.coli")
barplot(prot_composition_mycobacterium,main="Mycobacterium")

dna_composition_e.coli/sum(dna_composition_e.coli)
dna_composition_mycobacterium/sum(dna_composition_mycobacterium)

prot_composition_e.coli/sum(prot_composition_e.coli )
prot_composition_mycobacterium/sum(prot_composition_mycobacterium)

```
## This chunk contains code for:
1. Creating codon usage table
2. Creating codon usage chart
3. Codon usage bias showing in a data frame for a more organised look
4. Creating RSCU chart

```{r}
codons <- c("aaa", "aac", "aag", "aat", "aca", "acc", "acg", "act", "aga", "agc", "agg", "agt", 
            "ata", "atc", "atg", "att", "caa", "cac", "cag", "cat", "cca", "ccc", "ccg", "cct", 
            "cga", "cgc", "cgg", "cgt", "cta", "ctc", "ctg", "ctt", "gaa", "gac", "gag", "gat", 
            "gca", "gcc", "gcg", "gct", "gga", "ggc", "ggg", "ggt", "gta", "gtc", "gtg", "gtt", 
            "taa", "tac", "tag", "tat", "tca", "tcc", "tcg", "tct", "tga", "tgc", "tgg", "tgt", 
            "tta", "ttc", "ttg", "ttt")

frequencies.e <- c(44592, 28454, 13521, 23049, 9116, 31139, 19081, 11689, 2573, 21291, 1420, 11487, 
                 5486, 33524, 37007, 40501, 20402, 12890, 38397, 17105, 11163, 7238, 31074, 9225, 
                 4619, 29441, 7079, 27979, 5149, 14811, 70714, 14586, 52679, 25347, 23659, 42449, 
                 26743, 34117, 45082, 20185, 10350, 39536, 14581, 32779, 14430, 20350, 34996, 24213, 
                 2726, 16160, 294, 21241, 9303, 11390, 11830, 11111, 1219, 8574, 20196, 6802, 18323, 
                 21974, 18148, 29587)
frequencies.m <- c(5914, 19453, 16498, 5402, 5107, 35804, 16970, 3950, 1811, 16505, 4899, 4031, 
                 2656, 34959, 20467, 6822, 8377, 17186, 23783, 6908, 6952, 17902, 33509, 3570, 
                 8205, 30335, 25506, 8961, 4562, 18102, 47704, 5878, 16908, 42262, 30500, 16393, 
                 13047, 56598, 47821, 10896, 9956, 43524, 17592, 17626, 5494, 34454, 40998, 8591, 
                 664, 14580, 1613, 5882, 4151, 12686, 21736, 2494, 2464, 8503, 14971, 2645, 1728, 
                 23468, 19257, 6194)

codon_table<- data.frame(Codon = codons, FrequencyE.coli = frequencies.e, FrecuencyMycobacterium = frequencies.m)

print(codon_table)


par(mfrow = c(1, 2))

barplot(frequencies.e, 
        main = "Codon Usage in E. coli", 
        names.arg = codons, 
        las = 2, 
        col = rainbow(length(frequencies.e)), 
        ylab = "Frequency")

barplot(frequencies.m, 
        main = "Codon Usage in Mycobacterium", 
        names.arg = codons, 
        las = 2, 
        col = rainbow(length(frequencies.m)), 
        ylab = "Frequency")

RSCU_E.coli <- uco(E.coliDNA,index="rscu",as.data.frame=TRUE)
RSCU_Mycobacterium <- uco(MycobacteriumDNA,index="rscu",as.data.frame=TRUE)
RSCU_E.coli
RSCU_Mycobacterium


par(mfrow=c(1,2))
 barplot(RSCU_E.coli$RSCU,main="RSCUinE.coli",
 las=2,names.arg = codons,col=rainbow(length(RSCU_E.coli)),ylab="RSCU")
 barplot(RSCU_Mycobacterium$RSCU,main="RSCUinMycobacterium",
 las=2,names.arg = codons,col=rainbow(length(RSCU_Mycobacterium)),ylab="RSCU")

```

